vision-documents/variadic-generics.md

```diff
@@ -0,0 +1,515 @@
+    return List(rest...)
+  }
+}
+
+let list = List(1, "Hello", true)
+let firstRemoved = list.firstRemoved() // 'List("Hello", true)'
+```
+
+The body of `firstRemoved` decomposes `Element` into the components of its shape -- one value of type `First` and a value pack of type `Rest...` -- effectively removing the first element from the list.
+
+### Pack iteration
+
+All list operations can be expressed using pack expansion expressions by factoring code involving statements into a function or closure. However, this approach does not allow for short-circuiting, because the pattern expression will always be evaluated once for every element in the pack. Further, requiring a function or closure for code involving statements is unnatural. Allowing `for-in` loops to iterate over packs solves both of these problems.
+
+Value packs can be expanded into the source of a for-in loop, allowing you to iterate over each element in the pack and bind each value to a local variable:
+
+```swift
+func allEmpty<T...>(_ array: [T]...) -> Bool {
+  for array in array... {
+    guard array.isEmpty else { return false }
+  }
+  
+  return true
+}
+```
+
+The type of the local variable `array` in the above example is an `Array` of an opaque element type with the requirements that are written on `T`. For the *i*th iteration, the element type is the *i*th type parameter in the type parameter pack `T`.
+
+### Pack element projection
+
+Use cases for variadic generics that break up pack iteration across function calls, require random access, or operate over concrete packs are supported in this design by projecting individual elements out from a parameter pack. Because elements of the pack have different types, there are two approaches to pack element projection; using an `Int` index which will return the dynamic type of the element, and using a statically typed index which is parameterized over the requested pack element type.
+
+#### Dynamic pack indexing with `Int`
+
+Dynamic pack indexing is useful when the specific type of the element is not known, or when all indices must have the same type, such as for index manipulation or storing an index value. Packs will support subscript calls with an `Int` index, which will return the dynamic type of the pack element directly as the opened underlying type that can be assigned to a local variable with opaque type. Values of this type need to be erased or cast to another type to return an element value from the function:
+
+```swift
+func element<T...>(at index: Int, in t: T...) where T: P -> any P {
+  // The subscript returns 'some P', which is erased to 'any P'
+  // based on the function return type.
+  let value: some P = t[index]
+  return value
+}
+```
+
+Consider the following `ChainCollection` data structure, which has a type parameter pack of collections with the same `Element` type:
+
+```swift
+struct ChainCollection<Element, C...> where C: Collection, C.Element == Element {
+  var collection: C...
```

Comment:
```suggestion
  var collections: C...
```
(either that or the following should change to reference `collection` rather than `collections`.)

Reply:
Good catch! I think the reference should say `collection`; we're recommending the singular naming convention for packs